sdks/python/apache_beam/io/filesystem_test.py

```diff
@@ -0,0 +1,233 @@
+  # in compressed file and not just in the internal buffer
+  read_block_size = 4
+
+  def _create_compressed_file(self, compression_type, content,
+                              name='', suffix=''):
+    file_name = self._create_temp_file(name, suffix)
+
+    if compression_type == CompressionTypes.BZIP2:
+      compress_factory = bz2.BZ2File
+    elif compression_type == CompressionTypes.GZIP:
+      compress_factory = gzip.open
+    else:
+      assert False, "Invalid compression type: %s" % compression_type
+
+    with compress_factory(file_name, 'w') as f:
+      f.write(content)
+
+    return file_name
+
+  def test_seekable_enabled_on_read(self):
+    readable = CompressedFile(open(self._create_temp_file(), 'r'))
+    self.assertTrue(readable.seekable)
+
+  def test_seekable_disabled_on_write(self):
+    writeable = CompressedFile(open(self._create_temp_file(), 'w'))
+    self.assertFalse(writeable.seekable)
+
+  def test_seekable_disabled_on_append(self):
+    writeable = CompressedFile(open(self._create_temp_file(), 'a'))
+    self.assertFalse(writeable.seekable)
+
+  @parameterized.expand([CompressionTypes.BZIP2, CompressionTypes.GZIP])
+  def test_seek_set(self, compression_type):
+    file_name = self._create_compressed_file(compression_type, self.content)
+
+    compressed_fd = CompressedFile(open(file_name, 'r'), compression_type,
+                                   read_size=self.read_block_size)
+    reference_fd = StringIO(self.content)
+
+    random_position = randint(0, len(self.content) - 1)
+    compressed_fd.seek(random_position, os.SEEK_SET)
+    reference_fd.seek(random_position, os.SEEK_SET)
+
+    expected_position = random_position
+    uncompressed_position = compressed_fd.tell()
+    reference_position = reference_fd.tell()
+
+    self.assertEqual(uncompressed_position, expected_position)
+    self.assertEqual(reference_position, expected_position)
+
+    uncompressed_line = compressed_fd.readline()
+    reference_line = reference_fd.readline()
+
+    self.assertEqual(uncompressed_line, reference_line)
+
+  @parameterized.expand([CompressionTypes.BZIP2, CompressionTypes.GZIP])
+  def test_seek_cur(self, compression_type):
+    file_name = self._create_compressed_file(compression_type, self.content)
+
+    compressed_fd = CompressedFile(open(file_name, 'r'), compression_type,
+                                   read_size=self.read_block_size)
+    reference_fd = StringIO(self.content)
+
+    # skip to the middle without seeking
+    mid_point = len(self.content) / 2
+    _ = compressed_fd.read(mid_point)
+    _ = reference_fd.read(mid_point)
+
+    expected_position = mid_point
+    uncompressed_position = compressed_fd.tell()
+    reference_position = reference_fd.tell()
+
+    self.assertEqual(uncompressed_position, expected_position)
+    self.assertEqual(reference_position, expected_position)
+
+    seek_position = randint(-1 * mid_point, mid_point)
```

Comment:
Done. This exercise revealed an interesting behavior of cStringIO:
If we seek out of bound of a buffer through cStringIO then call tell() it will report
the out of bound position. If we do a read() or readline() prior to tell() then tell()
will report the correct position which is the size of the buffer.

CompressedFile's tell() never returns out of bound numbers. I've documented this difference in the testcase.